src/clients/UBAClient/UBAClientWithManual.ts

```diff
@@ -0,0 +1,119 @@
+import { BigNumber } from "ethers";
+import { RelayerFeeDetails } from "../../relayFeeCalculator";
+
+/**
+ * UBAClientState is a type that represents the state of a UBAClient.
+ */
+type UBAClientState = {
+  spoke: {
+    [chainId: number]: {
+      flows: UbaFlow[];
+      latestBlockNumber: number;
+      bundleEndBlocks: number[];
+      openingBalances: {
+        [token: string]: {
+          blockNumber: number;
+          balance: BigNumber;
+        }[];
+      };
+    };
+  };
+};
+
+/**
+ * UBAClient is a base class for UBA functionality. It provides a common interface for UBA functionality to be implemented on top of or extended.
+ * It is intended to only be used directly when RPC calls are __***not***__ required, e.g. in tests, Scraper API.
+ */
+export class UBAClientManual extends BaseUBAClient {
+  /**
+   * Constructs a UBAClient instance.
+   * @param chainIdIndices A list of available chain indices that should be reasoned about
+   * @param state Manual state that acts as the source of truth in this model. Note that these are initial values and can be modified by extended class features
+   * @param logger Optional logger to be used for logging
+   */
+  constructor(
+    readonly chainIdIndices: number[],
+    protected readonly state: UBAClientState,
+    readonly logger?: winston.Logger
+  ) {
+    super(chainIdIndices, logger);
+    assert(chainIdIndices.length > 0, "No chainIds provided");
+    assert(Object.values(state.spoke).length > 0, "No SpokePools provided");
+  }
+
+  protected resolveClosingBlockNumber(chainId: number, blockNumber: number): number {
+    return this.state.spoke[chainId].bundleEndBlocks.find((bundleEndBlock) => bundleEndBlock <= blockNumber) ?? -1;
+  }
+
+  public getOpeningBalance(
+    chainId: number,
+    spokePoolToken: string,
+    hubPoolBlockNumber?: number
+  ): OpeningBalanceReturnType {
+    const spoke = this.state.spoke[chainId];
+    const token = spoke.openingBalances[spokePoolToken];
+    if (!isDefined(token)) {
+      throw new Error(`Could not resolve ${chainId} token ${spokePoolToken} at block ${hubPoolBlockNumber}`);
+    }
+    const openingBalance = token
+      .slice()
+      .reverse()
+      .find((balance) => !hubPoolBlockNumber || balance.blockNumber <= hubPoolBlockNumber);
+    if (!isDefined(openingBalance)) {
+      throw new Error(`Could not resolve ${chainId} token ${spokePoolToken} at block ${hubPoolBlockNumber}`);
+    }
+    const { blockNumber, balance: spokePoolBalance } = openingBalance;
+    return { blockNumber, spokePoolBalance };
+  }
+
+  public getFlows(chainId: number, fromBlock?: number, toBlock?: number): UbaFlow[] {
+    const { flows } = this.state.spoke[chainId];
+    if (flows.length === 0) {
+      return [];
+    }
+    return flows.filter((flow) =>
+      isDefined(fromBlock)
+        ? flow.blockNumber >= fromBlock
+        : true && isDefined(toBlock)
```

Comment:
```suggestion
        : isDefined(toBlock)
```

why is that `true` there?